### DISCUSSION TITLE
Map not rendering correctly on load, but fine on viewport resize; forceful reload?

### DISCUSSION THREAD

**Discussion by ero-qt** [Leaflet/Leaflet#8795] [Q&A] [answered]
I'm creating some templated code which I plan on using in the future to create maps more easily: simply provide the tiles and the markers and that'd be all.

I'm running into an issue where, due to me adding things manually to the HTML `head` and `body`, the map does not render correctly when the page is first loaded (screenshots below). Once I resize the viewport in any way, the map reloads, displays fully and jumps into the correct position.

What could be the cause of this? Can I avoid it? Can I force this reload to happen manually?

Here's my code:
`index.html`: The HTML of one of my template implementations.
```html
<!DOCTYPE html>
<html lang="en">

<head>
  <script src="impl1.js" type="module"></script>

  <link rel="icon" href="icon.ico">
  <title>Template Implementation</title>
</head>

</html>
```
`impl1.js`: The JavaScript code of the same implementation.
```js
import * as G from 'global.js';

G.init();
G.render();
```
`global.js`: The templated JavaScript code, shared across implementations.
```js
import { } from 'https://code.jquery.com/jquery-3.6.3.min.js';
import { } from 'https://unpkg.com/leaflet@1.9.3/dist/leaflet.js';

export function init() {
  $('<div/>', {
    id: 'map'
  }).appendTo('body');

  $('<meta/>', {
    charset: 'utf-8'
  }).appendTo('head');

  $('<link/>', {
    rel: 'stylesheet',
    href: 'https://unpkg.com/leaflet@1.9.3/dist/leaflet.css'
  }).appendTo('head');

  $('<link/>', {
    rel: 'stylesheet',
    href: 'global.css'
  }).appendTo('head');
}

export function render() {
  const tileSize = 256;
  const bounds = [[0, 0], [-tileSize, tileSize]];

  const map = L.map("map", {
    crs: L.CRS.Simple,
    maxBounds: bounds,
    tileSize: tileSize,
    zoomDelta: .5,
    maxBoundsViscosity: 1
  }).setView([0, 0], 1);

  L.tileLayer("tiles/{z}/tile_{x}_{y}.png", {
    minZoom: 1,
    maxZoom: 5,
    updateInterval: 10,
    bounds: bounds
  }).addTo(map);
}
```

There are no error messages. It just doesn't do what I want.

---

### First load:
![Image: Map on first load of the web page](https://user-images.githubusercontent.com/56401411/212326423-01dd2f6c-0771-4b90-8689-1cb762e38ed6.png)

### After resize:
![Image: Map after resize of the web page](https://user-images.githubusercontent.com/56401411/212326525-40264b0d-5855-41ce-b16d-c6b846bad129.png)

**Comment by IvanSanchez** [comment]
Classic missing `invalidateSize()`. Shouldn't happen anymore once https://github.com/Leaflet/Leaflet/pull/8612 gets into a release.

**Reply by ero-qt** [reply]
This does not fix it. As the documentation states, that function only applies any changes when the container size changed, which it didn't.

**Reply by IvanSanchez** [reply]
You sure? I'm not seeing any CSS rules in your code that give height to the map container.

**Reply by ero-qt** [reply]
`global.css` (which I also add to `head` in the `global.js` script) looks like this:
```css
html,
body {
  height: 100%;
  margin: 0
}

.leaflet-container {
  background: #6767b4;
  outline: 0
}

#map {
  height: 100%
}
```
I didn't deem this as an issue, since it's worked before making it dynamic like this, and since it does work when the viewport changes size.

**Reply by IvanSanchez** [reply]
Try doing a `console.log( document.getElementById('map').getBoundingRect().height );` once during initialization, and then once after the document has settled. I bet they're gonna have different values.

Making things dynamic has the nasty habit of keeping stuff half-initialized until the document flow is recalculated (or any other similar reason out of a half dozen).

The usual advice of using viewport units (`vh`/`vw`) instead of percentages applies here too.

**Reply by ero-qt** [reply]
Using
```js
$('<div/>', {
  id: 'map',
  height: '100vh'
}).appendTo('body');
```
Seems to have done the trick. Is there something I should be aware of when setting the height through that? Or will this be fine?

**Comment by ero-qt** [comment, marked as answer]
https://github.com/Leaflet/Leaflet/discussions/8795#discussioncomment-4678615
